httpbin/handlers.go

```diff
@@ -1016,8 +1015,18 @@ func (h *HTTPBin) Bearer(w http.ResponseWriter, r *http.Request) {
 
 // Hostname - returns the hostname.
 func (h *HTTPBin) Hostname(w http.ResponseWriter, r *http.Request) {
-	body, _ := json.Marshal(hostnameResponse{
+	body, _ := jsonMarshalNoEscape(hostnameResponse{
 		Hostname: h.hostname,
 	})
 	writeJSON(w, body, http.StatusOK)
 }
+
+// json.Marshal escapes HTML in strings while httpbin does not, so
+// we need to set up the encoder manually to reproduce that behavior.
+func jsonMarshalNoEscape(value interface{}) ([]byte, error) {
+	buffer := &bytes.Buffer{}
+	encoder := json.NewEncoder(buffer)
+	encoder.SetEscapeHTML(false)
```

Comment:
I'm not sure I understand what problem this is solving.  AFAICT, go-httpbin is already returning unescaped HTML in the response:

```python
>>> import requests
>>> requests.post('https://httpbingo.org/post', data="<html><body><h1>hello world</h1></body></html>", headers={"Content-Type": "text/html"}).json()["data"]
'<html><body><h1>hello world</h1></body></html>'
```

```bash
$ curl -sSL -X POST -d "<html><body><h1>hello world</h1></body></html>" -H "Content-Type: text/html" https://httpbingo.org/post | jq -r .data
<html><body><h1>hello world</h1></body></html>
```

Looking at the "raw" output from curl might shed a little light here, as it seems the `<` and `>` characters in the data are encoded as unicode references:

```
"data":"\u003chtml\u003e\u003cbody\u003e\u003ch1\u003ehello world\u003c/h1\u003e\u003c/body\u003e\u003c/html\u003e"
```

But, as you can see in the python and jq examples above, this form and the unencoded form are semantically identical.  What tool are you using that can't handle this form?

(Please let me know if I'm missing some nuance here!)

Reply:
Never mind, having thought about this for a few more minutes, I think I understand the issue:  Ideally, we should just send back the incoming data byte-for-byte, where the current approach does change it.

But … I'm not sure how to reconcile that with the fact that it must be encoded as valid JSON, so there is always going to be some input data that will be transformed in the output.  In the simplest case, that might just mean escaping double quotes in ASCII text, but it gets much more complex.  E.g. let's look at what happens when we post binary data (in this case a 1x1 png image).

Original httpbin returns it as a base64-encoded `data:` URL:

```
$ curl -sSL --data-binary @tiny.png -H "Content-Type: image/png"  https://httpbin.org/post | jq .data
"data:application/octet-stream;base64,iVBORw0KGgoAAAANSUhEUgAAAAEAAAABCAAAAAA6fptVAAAACklEQVQIW2O4AwAA3gDddw4iTwAAAABJRU5ErkJggg=="
```

go-httpbin returns it basically as-is, accounting for encoding the bytes in JSON-compatible format:

```
$ curl -sSL --data-binary @tiny.png -H "Content-Type: image/png"  https://httpbingo.org/post | jq .data
"�PNG\r\n\u001a\n\u0000\u0000\u0000\rIHDR\u0000\u0000\u0000\u0001\u0000\u0000\u0000\u0001\b\u0000\u0000\u0000\u0000:~�U\u0000\u0000\u0000\nIDAT\b[c�\u0003\u0000\u0000�\u0000�w\u000e\"O\u0000\u0000\u0000\u0000IEND�B`�"
```

So, basically, I'm not sure what the right call is here.

Reply:
Okay, I'm back to being in favor of this change, after actually reading [the documentation for SetEscapeHTML](https://pkg.go.dev/encoding/json#Encoder.SetEscapeHTML):

> SetEscapeHTML specifies whether problematic HTML characters should be escaped inside JSON quoted strings. The default behavior is to escape &, <, and > to \u0026, \u003c, and \u003e to avoid certain safety problems that can arise when embedding JSON in HTML. 

I was initially worried that it would be doing something more invasive than this.  Seems like a reasonable update to me.

Reply:
Thanks - yup both forms are valid JSON, but unescaped allows the endpoint to be something like an echo server, which is how coreruleset tests are using it

Reply:
> yup both forms are valid JSON, but unescaped allows the endpoint to be something like an echo server, which is how coreruleset tests are using it

I'm still fine with this change, but I'd suggest that if coreruleset is comparing JSON output byte-for-byte then it is far more brittle than it should be.

Will changes in the formatting of the JSON, like pretty-printing, break this in the future?